schemas: add glyph.fbs and plist.fbs

anthrotype opened the pull request:
The Glyph type mimics the properties of a defcon.Glyph or ufoLib2.Glyph (rather than literally following the UFO spec), to aid interoperability.
E.g. there's no `Advance` table with a `width` and `height` attributes, but the latter are directly attributes of Glyph itself. The `unicodes` are list of integers, not hex strings (as in the UFO XML serialization).
Plist is weird because apparently one can't have vector of unions or unions of anything other than a table (structs or literals not allowed) in languages other than c++, so one must add further levels of indirection.

anthrotype commented on schemas/ufo/glyph.fbs:
```diff
@@ -0,0 +1,60 @@
+include "fontinfo.fbs";
+include "plist.fbs";
+
+//namespace FlatFont.Ufo;
```
I left this commented out because fontinfo.fbs has it that way, but for python at least we really want this namespaced, because one does not really want to do a blanket `import Glyph`

anthrotype replied:
@behdad so WDYT about the namespace? I think it's a good idea to have one.

behdad replied:
@anthrotype It's a must. I added it. @rsheeter commented it out because he was lazy figuring out how to do it in his rust code I suppose.

behdad replied:
So yes, uncomment it and let's fix the rust code.

behdad commented on schemas/ufo/glyph.fbs:
```diff
@@ -0,0 +1,60 @@
+// we include fontinfo.fbs this because the Guideline table is already defined there
+include "fontinfo.fbs";
```
Let's move the comment above to the line of include and be: `` // for Guideline``

behdad replied:
Also, maybe move Guideline into separate file and include it from both.

behdad replied:
Perhaps into a separate `types.fbs` or `base.fbs` or something simple.

anthrotype replied:
added base.fbs and moved Guideline

behdad commented on schemas/ufo/glyph.fbs:
```diff
@@ -0,0 +1,60 @@
+  guidelines: [Guideline];
+}
+
+table Image {
```
Comments with URLs of these tables need to be added, eg.:
```
/// https://unifiedfontobject.org/versions/ufo3/glyphs/glif/#image
```

behdad reviewed (commented):
LGTM! We should add a build system (which one!) and add tests.

behdad commented on schemas/ufo/glyph.fbs:
```diff
@@ -0,0 +1,60 @@
+}
+
+struct Transform {
+  values: [float:6];
```
Ideally a union with `xScale, xyScale, yxScale, yScale, xOffset, yOffset`. But union can't take struct?

I feel like `Transform` should also live in `base.fbs`. Perhaps in a toplevel `base.fbs` even, not in `ufo/base.fbs` even.

anthrotype replied:
why union? We can have Transform struct use names for each field instead of a fixed-size array, but in defcon/ufoLib2 APIs it is a 6-tuple, not a dictionary.

Yes I can move to base.fbs if you prefer, though it's only used in glyph.

behdad replied:
> why union?

Such that we can use it both as named or array. For now maybe make it named.

> Yes I can move to base.fbs if you prefer, though it's only used in glyph.

Sure keep it here for now. Thanks!

anthrotype replied:
Do you prefer like this?

```
namespace FlatFont;

struct Transform {
  xScale: float;
  xyScale: float;
  yxScale: float;
  yScale: float;
  xOffset: float;
  yOffset: float;
}
```

Note that there's been a lot of confusions about these symbolic names for the 2x3 affine fields. In FontTools as well as UFO specification and a lot of the lib that derive from those, the _names_ of `xyScale` and `yxScale` are swapped compared to similar structs found in FreeType and Cairo, while they actually mean the same thing, see: 
https://github.com/fonttools/fonttools/blob/445108f735b22a5ca37f669808d47906d024fe24/Lib/fontTools/ttLib/tables/otData.py#L1633-L1648

Would be best to just avoid the names and use a fixed length array of 6 floats IMO.

behdad replied:
Even if you use 6 floats the order needs to be specified, so you are not actually *solving* the problem.

I prefer the spelled out.

Also, set the default for `xScale` and `yScale` to 1.0. How about that?

behdad replied:
Anyway, we can reconsider later.

anthrotype replied:
you _cannot_ set defaults on structs.. should I turn Transform into a `table` to set their defaults?

How shall I call xyScale and yxScale? using fontTools/UFO convention of FreeType/Cairo?

anthrotype replied:
(I mean't one cannot set defaults on structs)

behdad replied:
Yeah let's forget about struct and make it table, with defaults, which would make it stored more optimized.

The naming should be `xyScale` first and `yxScale` after indeed.

anthrotype replied:
> The naming should be xyScale first and yxScale after indeed.

funny you say that.. in COLRv1 Affine2x3 we ended up following the FreeType/Cairo naming and have yxScale _before_ xyScale.
(Of course the actual ordering is the same, only the label differ, which adds to the confusion).

anthrotype replied:
we call "yxScale" the "y-part of the x basis vector" and xyScale "the x-part of the y basis vector" (following 3blue1brown's definitions) /cc @rsheeter

behdad replied:
> funny you say that..

Not funny actually that groupthink prevailed...

> we call "yxScale" the "y-part of the x basis vector"

That's English, which is *not* universal. It should be "xyScale", because math / computer code is:
```
x' = xxScale * x + xyScale * y
y' = yxScale * x + yyScale * y
```

behdad commented:
LGTM recent changes. Thanks!

anthrotype commented:
I changed Transform to a table with defaults, moved both Transform and Guideline to schemas/ufo/misc.fbs (for all residual stuff).
I also moved plist.fbs one level up to schemas/ directory (and `FlatFont` namespace) because it is not ufo-only (designspace also uses plist for the lib element).
Note that all schemas/ufo/*.fbs share the same "namespace FlatFont.Ufo" now. I'm not sure if we need to give each *.fbs file its unique scope (e.g. `FontInfo.Ufo.Glyph`, etc.) or if they can/should share the same namespace.
For the last commit were I have updated the rust generated code, I only rebuilt the fontinfo.fbs (and misc.fbs which is included by the latter). It exploded from one single .rs file to a directory tree with various submodules, one for each table I think.
Cargo build passes so it seems to be working, but please double check I did it correctly.

For some reasons, I noticed that the order in which I pass the *.fbs to the `flatc --rust` compiler changes the output of the generated top-level `mod.rs` file: i.e. I have to say `flatc --rust -o ufoff/src/flatbuffers_generated schemas/ufo/{misc,fontinfo}.fbs` to produce the expected result, if I change the order and pass fontinfo.fbs and then misc.fbs, the latter seems to overwrite the fontinfo related stuff in the top-level mod.rs... I have no idea why.

behdad commented:
> Note that all schemas/ufo/*.fbs share the same "namespace FlatFont.Ufo" now. I'm not sure if we need to give each *.fbs file its unique scope (e.g. `FontInfo.Ufo.Glyph`, etc.) or if they can/should share the same namespace.

No I think the namespace should follow the directory structure indeed.

anthrotype commented:
The ufoff README instructions on how to regenerate the rust flatbuffers is outdated now that we added namespaces, it should read something like

flatc --rust -o ufoff/src/flatbuffers_generated schemas/**/*.fbs

however the order of the arguments seems to matter in an unclear way, as I noted above. Doing the glob will list files alphabetically on my shell, and misc comes after fontinfo, which wipes the fontinfo stuff from the top-level mod.rs...

I'll leave it to you to fix, enough fontbuffers for today.

anthrotype commented:
it seems like we hit this https://github.com/google/flatbuffers/issues/6800 -- but I don't understand what the workaround is

rsheeter commented:
@behdad could you elaborate on the objection to the naming y part of x basis? - it was the sanest scheme I came across. Most others seemed to simply ignore what the fields _mean_ entirely.

behdad commented:
> @behdad could you elaborate on the objection to the naming y part of x basis? - it was the sanest scheme I came across. Most others seemed to simply ignore what the fields _mean_ entirely.

As I said, it should be named in math, /not/ English, order:
```
x' = xxScale * x + xyScale * y
y' = yxScale * x + yyScale * y
```

rsheeter commented:
But the math doesn't care what names you use?
